File: frontend/src/api/preview.ts
import { request } from './config'

export interface TextbookOption {
  subject: string
  grade: number
  semester: string
  version: string
  label: string
  source_name: string
  source_url: string
  status: string
}

export interface PreviewUnit {
  unit: number
  title: string
  total_items: number
  completed_items: number
  source_status: string
}

export interface PreviewItem {
  item_key: string
  item_type: string
  category_label: string
  word: string
  pronunciation: string
  meaning: string
  hint: string
  sample_sentences: string[]
  practice_prompt: string
  completed: boolean
}

export interface PreviewUnitDetail {
  subject: string
  grade: number
  semester: string
  textbook_version: string
  unit: number
  title: string
  source_name: string
  source_url: string
  source_status: string
  guidance: string
  items: PreviewItem[]
}

export function getTextbooks(): Promise<{ items: TextbookOption[] }> {
  return request<{ items: TextbookOption[] }>({ url: '/preview/textbooks', method: 'GET' })
}

export function getPreviewUnits(params: {
  subject: string
  grade: number
  semester: string
  textbook_version: string
}): Promise<{ subject: string; grade: number; semester: string; textbook_version: string; units: PreviewUnit[] }> {
  const q = encodeParams(params)
  return request<{ subject: string; grade: number; semester: string; textbook_version: string; units: PreviewUnit[] }>({
    url: `/preview/units?${q}`,
    method: 'GET',
  })
}

export function getPreviewUnitDetail(params: {
  subject: string
  grade: number
  semester: string
  textbook_version: string
  unit: number
}): Promise<PreviewUnitDetail> {
  const { unit, ...rest } = params
  return request<PreviewUnitDetail>({ url: `/preview/units/${unit}?${encodeParams(rest)}`, method: 'GET' })
}

export function completePreviewItem(data: {
  subject: string
  grade: number
  semester: string
  textbook_version: string
  unit: number
  item_key: string
  item_type: string
}): Promise<{ item_key: string; completed: boolean; completed_at: string }> {
  return request<{ item_key: string; completed: boolean; completed_at: string }>({
    url: '/preview/complete',
    method: 'POST',
    data,
  })
}

export interface ChallengeOption {
  text: string
  is_correct: boolean
  index: number
}

export interface ChallengeQuestion {
  word: string
  pinyin: string
  type: string
  question_text: string
  options: ChallengeOption[]
  correct_answer: string
}

export function getUnitChallenge(params: {
  subject: string
  grade: number
  semester: string
  unit_no: number
}): Promise<{ subject: string; grade: number; semester: string; unit: number; questions: ChallengeQuestion[] }> {
  const q = encodeParams(params)
  return request<{ subject: string; grade: number; semester: string; unit: number; questions: ChallengeQuestion[] }>({
    url: `/preview/unit-challenge?${q}`,
    method: 'GET',
  })
}

export interface ChallengeResult {
  points_earned: number
  correct_count: number
  total: number
  total_points: number
  streak_days: number
  words_mastered: number
  new_badges: { id: string; name: string; desc: string }[]
}

export function submitChallengeResult(data: {
  subject: string
  grade: number
  results: { word: string; correct: boolean }[]
}): Promise<ChallengeResult> {
  return request<ChallengeResult>({ url: '/preview/challenge-result', method: 'POST', data })
}

export interface StudiedUnit {
  subject: string
  grade: number
  semester: string
  unit: number
  title: string
  completed_items: number
  total_items: number
  percent: number
}

export interface ParentSummary {
  period_days: number
  weekly_completed: number
  subject_breakdown: Record<string, number>
  studied_units: StudiedUnit[]
  review_suggestions: string[]
}

export function getParentSummary(params: { days?: number; grade: number }): Promise<ParentSummary> {
  const q = encodeParams({ days: params.days ?? 7, grade: params.grade })
  return request<ParentSummary>({ url: `/preview/parent-summary?${q}`, method: 'GET' })
}

export function explainPreviewItem(data: {
  subject: string
  grade: number
  word: string
  item_type: string
  category_label?: string
  unit_title?: string
  meaning?: string
}): Promise<{ word: string; explanation: string }> {
  return request<{ word: string; explanation: string }>({
    url: '/preview/explain',
    method: 'POST',
    data,
  })
}

function encodeParams(params: Record<string, string | number>) {
  return Object.entries(params)
    .map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(String(value))}`)
    .join('&')
}

File: backend/app/schemas/preview.py
from datetime import datetime

from pydantic import BaseModel


class TextbookOption(BaseModel):
    subject: str
    grade: int
    semester: str
    version: str
    label: str
    source_name: str
    source_url: str
    status: str


class TextbookOptionsResponse(BaseModel):
    items: list[TextbookOption]


class PreviewUnitItem(BaseModel):
    unit: int
    title: str
    total_items: int
    completed_items: int
    source_status: str


class PreviewUnitsResponse(BaseModel):
    subject: str
    grade: int
    semester: str
    textbook_version: str
    units: list[PreviewUnitItem]


class PreviewItem(BaseModel):
    item_key: str
    item_type: str
    category_label: str
    word: str
    pronunciation: str = ""
    meaning: str
    hint: str = ""
    sample_sentences: list[str] = []
    practice_prompt: str
    completed: bool = False


class PreviewUnitDetailResponse(BaseModel):
    subject: str
    grade: int
    semester: str
    textbook_version: str
    unit: int
    title: str
    source_name: str
    source_url: str
    source_status: str
    guidance: str
    items: list[PreviewItem]


class CompletePreviewItemRequest(BaseModel):
    subject: str
    grade: int
    semester: str
    textbook_version: str
    unit: int
    item_key: str
    item_type: str


class CompletePreviewItemResponse(BaseModel):
    item_key: str
    completed: bool
    completed_at: datetime


class ExplainPreviewItemRequest(BaseModel):
    subject: str
    grade: int
    word: str
    item_type: str
    category_label: str = ""
    unit_title: str = ""
    meaning: str = ""


class ExplainPreviewItemResponse(BaseModel):
    word: str
    explanation: str


class StudiedUnit(BaseModel):
    subject: str
    grade: int
    semester: str
    unit: int
    title: str
    completed_items: int
    total_items: int
    percent: int


class ParentSummaryResponse(BaseModel):
    period_days: int
    weekly_completed: int
    subject_breakdown: dict[str, int]
    studied_units: list[StudiedUnit]
    review_suggestions: list[str]


class ChallengeOption(BaseModel):
    text: str
    is_correct: bool
    index: int


class ChallengeQuestion(BaseModel):
    word: str
    pinyin: str = ""
    type: str
    question_text: str
    options: list[ChallengeOption]
    correct_answer: str


class UnitChallengeResponse(BaseModel):
    subject: str
    grade: int
    semester: str
    unit: int
    questions: list[ChallengeQuestion]


class ChallengeResultItem(BaseModel):
    word: str
    correct: bool


class ChallengeResultRequest(BaseModel):
    subject: str
    grade: int
    results: list[ChallengeResultItem]


class BadgeInfo(BaseModel):
    id: str
    name: str
    desc: str


class ChallengeResultResponse(BaseModel):
    points_earned: int
    correct_count: int
    total: int
    total_points: int
    streak_days: int
    words_mastered: int
    new_badges: list[BadgeInfo]
